src/gateway/Gateway.sol

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.13;
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {IERC20Permit} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Permit.sol";
+import {Pausable} from "@openzeppelin/contracts/security/Pausable.sol";
+
+/// @title ECG Gateway
+/// @notice Gateway to interract via multicall with the ECG
+/// Owner can select which calls are allowed
+abstract contract Gateway is Ownable, Pausable {
+    /// @notice Emitted when an external call fails with an error.
+    error CallExternalError(bytes innerError);
+
+    // never allow transferFrom to be whitelisted
+    // this avoid human error
+    bytes4 public immutable transferFromSelector =
+        bytes4(keccak256("transferFrom(address,address,uint256)"));
+
+    /// @notice emitted when a call is allowed or not by the function allowCall
+    event CallAllowed(
+        address indexed target,
+        bytes4 functionSelector,
+        bool isAllowed
+    );
+
+    /// @notice Stores calls to be executed after receiving a Balancer flash loan.
+    /// @dev The StoredCalls should only be set in the 'multicallWithBalancerFlashLoan' function
+    ///      Which is "onlyOwner" so the StoredCalls can only be set by the owner of the contract
+    bytes[] internal _storedCalls;
```

Comment:
This can move to `GatewayV1` because it's only needed for flashloans, not in the base features